### PR TITLE
Add user active stage check and enhance logging functionality

### PR DESCRIPTION
**Pull request by tomerp20**
This PR contains initial setup for monitoring and logging login for this user events:
Registration 
Login
Logout
Start Stage
End Stage - fail: wrong flag
End Stage - success
Rate limit Exceeding on stage solve attempt

The PR reference's to this trello ticket:
https://trello.com/c/xhrH0K8K/81-track-users-event-from-ctfd-into-opensearch

### CODE REVIEW

**Comment by amit-sha-webiks** [comment]
looking GOOD :)

File: CTFd/api/v1/challenges.py [hunk 1]
```diff
@@ -582,18 +583,16 @@ def post(self):
         # Anti-bruteforce / submitting Flags too quickly
         kpm = current_user.get_wrong_submissions_per_minute(user.account_id)
         kpm_limit = int(get_config("incorrect_submissions_per_min", default=10))
+        submission = request_data.get("submission", "").encode("utf-8")
         if kpm > kpm_limit:
             if ctftime():
                 chal_class.fail(
                     user=user, team=team, challenge=challenge, request=request
                 )
             log(
```
Comment (shacharmo) [inline CTFd/api/v1/challenges.py:592]: Should use logger with log level, logger instance should also be be per module (in this case `challenges`)

File: CTFd/utils/initialization/__init__.py [hunk 2]
```diff
@@ -132,54 +132,63 @@ def init_template_globals(app):
...
-            logs["registrations"], maxBytes=10485760, backupCount=5
-        )
-
-        logger_submissions.addHandler(submission_log)
-        logger_logins.addHandler(login_log)
-        logger_registrations.addHandler(registration_log)
-    except IOError:
-        pass
+    LOG_FORMAT = (
+        "%(asctime)s|CTFd|%(levelname)s|%(message)s|%(event)s|extra_data=%(extra_data)s|trace_id=%(trace_id)s"
```
Comment (shacharmo) [inline CTFd/utils/initialization/__init__.py:136]: CTFd shouldn't be hardcoded, we already know were are in the context of CTFd, instead it should be the module which created the log

File: CTFd/utils/user/user_manager.py [hunk 3]
```diff
@@ -13,10 +13,14 @@ def __init__(self, user_id, access_token):
         idp_user = self.idp_user_instance.get_user()['data']
 
         # Assign attributes from idp_user
-        idp_fields = ['name', 'email', 'type', 'secret', 'website', 'affiliation', 'country', 'hidden', 'banned', 'verified', 'language']
+        idp_fields = ['name', 'email','custom:active_c' 'type', 'secret', 'website', 'affiliation', 'country', 'hidden', 'banned', 'verified', 'language']
```
Comment (shacharmo) [inline CTFd/utils/user/user_manager.py:16]: Use a better name like `current_challenge` for clarity

File: CTFd/utils/logging/__init__.py [hunk 4]
```diff
@@ -1,19 +1,35 @@
...
+        "extra_data": extra_data,
+        "trace_id": 123123,
+        "event": event,
+        
     }
-    props.update(kwargs)
-    msg = format.format(**props)
-    logger.info(msg)
+
+    print('now logging')
```
Comment (shacharmo) [inline CTFd/utils/logging/__init__.py:30]: Remove

File: CTFd/utils/security/auth.py [hunk 5]
```diff
@@ -23,10 +23,12 @@ def save_token(token):
     # Save the token in the database
     session['tokens']  = token
     return {'success': True, 'message': 'Token saved' }
+def replaceCognitoSubInUserId(sub):
```
Comment (shacharmo) [inline CTFd/utils/security/auth.py:26]: Isn't there a linter defined? There should be at least one empty row before this line

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: CTFd/api/v1/challenges.py [hunk 6]
```diff
@@ -582,18 +583,16 @@ def post(self):
...
-                "submissions",
-                "[{date}] {name} submitted {submission} on {challenge_id} with kpm {kpm} [TOO FAST]",
-                name=user.name,
-                submission=request_data.get("submission", "").encode("utf-8"),
-                challenge_id=challenge_id,
-                kpm=kpm,
+                "rate_limit_exceeded_in_solve_attempt",
+                " {user.name} submitted {submission} on {challenge_id} with kpm {kpm} [TOO FAST]",
+                submission = submission,
```
[the remark this answers is not in the document]
Comment (tomerp20) [reply CTFd/api/v1/challenges.py:595]: the stage (challange) id is added inside the log() function.
kpm added.

File: CTFd/utils/logging/__init__.py [hunk 7]
```diff
@@ -1,19 +1,35 @@
...
+    }
+    if user:
+        if hasattr(user, 'active_c'):
+            extra_data['stage_id'] = user.active_c
+        if hasattr(user, 'id'):
+            extra_data['user_id'] = user.id
+    
+    extra = {
+        "extra_data": extra_data,
+        "trace_id": 123123,
```
[the remark this answers is not in the document]
Comment (tomerp20) [reply CTFd/utils/logging/__init__.py:25]: Yes, the trace_id will be there in the next pr (:

File: CTFd/api/v1/challenges.py [hunk 8]
```diff
@@ -637,12 +636,9 @@ def post(self):
                     clear_challenges()
 
                 log(
-                    "submissions",
-                    "[{date}] {name} submitted {submission} on {challenge_id} with kpm {kpm} [CORRECT]",
-                    name=user.name,
-                    submission=request_data.get("submission", "").encode("utf-8"),
-                    challenge_id=challenge_id,
-                    kpm=kpm,
+                    "correct_stage_solve",
```
[the remark this answers is not in the document]
Comment (tomerp20) [reply CTFd/api/v1/challenges.py:640]: added kmp

File: CTFd/api/v1/challenges.py (hunk 1 above)
Comment (tomerp20) [reply CTFd/api/v1/challenges.py:592]: fixed

File: CTFd/utils/initialization/__init__.py [hunk 9]
```diff
@@ -132,54 +132,63 @@ def init_template_globals(app):
...
+    )
+    DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+    formatter = logging.Formatter(LOG_FORMAT, datefmt=DATE_FORMAT)
+    formatter.converter = time.gmtime  # Use UTC for timestamp
+
+    # logger_submissions = logging.getLogger("submissions")
+    # logger_logins = logging.getLogger("logins")
```
[the remark this answers is not in the document]
Comment (tomerp20) [reply CTFd/utils/initialization/__init__.py]: fixed

**Pull request merged**